extensions/cronet/src/main/java/com/google/android/exoplayer2/ext/cronet/CronetDataSource.java

```diff
@@ -568,6 +581,43 @@ public int read(byte[] buffer, int offset, int readLength) throws HttpDataSource
   public int read(ByteBuffer buffer) throws HttpDataSourceException {
     Assertions.checkState(opened);
 
+    return readIntoByteBuffer(buffer);
+  }
+
+  @Override
+  public synchronized void close() {
+    if (currentUrlRequest != null) {
+      currentUrlRequest.cancel();
+      currentUrlRequest = null;
+    }
+    if (readBuffer != null) {
+      readBuffer.limit(0);
+    }
+    currentDataSpec = null;
+    responseInfo = null;
+    exception = null;
+    finished = false;
+    if (opened) {
+      opened = false;
+      transferEnded();
+    }
+  }
+
+  /** Returns current {@link UrlRequest}. May be null if the data source is not opened. */
+  @Nullable
+  protected UrlRequest getCurrentUrlRequest() {
+    return currentUrlRequest;
+  }
+
+  /** Returns current {@link UrlResponseInfo}. May be null if the data source is not opened. */
+  @Nullable
+  protected UrlResponseInfo getCurrentUrlResponseInfo() {
+    return responseInfo;
+  }
+
+  // Internal methods.
+
+  private int readIntoByteBuffer(ByteBuffer buffer) throws HttpDataSourceException {
```

Comment:
This diff kind of sucks.

`close()` has not changed at all. The change here is that there is a new private method called `readIntoByteBuffer(ByteBuffer buffer)`, which is used in two places: `read(ByteBuffer buffer)` and in `open()` (to read the error response body).